action/action.go

```diff
@@ -17,15 +21,160 @@ import (
 // stateful is there just to make callers of opFromClaims more readable
 const stateful = false
 
-// Action describes one of the primary actions that can be executed in CNAB.
-//
-// The actions are:
-// - install
-// - upgrade
-// - uninstall
-type Action interface {
-	// Run an action, and record the status in the given claim
-	Run(*claim.Claim, credentials.Set, ...OperationConfigFunc) error
+// Action executes a bundle operation and helps save the results.
+type Action struct {
+	Claims         claim.Provider
+	Driver         driver.Driver
+	SaveAllOutputs bool
+	SaveOutputs    []string
+	SaveLogs       bool
+}
+
+// New creates an Action.
+// - Driver to execute the operation with.
+// - Claim Provider for persisting the claim data. If not set, calls to Save* will error.
```

Comment:
I believe that DataDog uses their own optimized persistence layer and I didn't want to prevent them from using the action package anymore. Adding @jlegrone to confirm.

This is something that would be great to collaborate on in a follow-up and figure out where we need to have compromises like this and where we can improve the abstractions so that everyone is happy using them. But for the first go at it, I kept things open "minded" to avoid problems with known implementations.